src/android/FileUtils.java

```diff
@@ -433,28 +434,41 @@ public void run(JSONArray args) throws JSONException, IOException {
         else if (action.equals("getDirectory")) {
             threadhelper( new FileOp( ){
                 public void run(JSONArray args) throws FileExistsException, IOException, TypeMismatchException, EncodingException, JSONException {
-                    String dirname=args.getString(0);
-                    String path=args.getString(1);
-                    JSONObject obj = getFile(dirname, path, args.optJSONObject(2), true);
-                    callbackContext.success(obj);
+                    String dirname = args.getString(0);
+                    String path = args.getString(1);
+                    String nativeURL = resolveLocalFileSystemURI(dirname).getString("nativeURL");
+                    boolean containsCreate = args.toString().contains("\"create\":true");
+
+                    if(containsCreate && needPermission(nativeURL, WRITE)) {
+                        getPermissionDir(WRITE);
+                    }
+                    else if(!containsCreate && needPermission(nativeURL, READ)) {
+                        getPermissionDir(READ);
+                    }
+                    else {
+                        JSONObject obj = getFile(dirname, path, args.optJSONObject(2), true);
+                        callbackContext.success(obj);
+                    }
                 }
             }, rawArgs, callbackContext);
         }
         else if (action.equals("getFile")) {
             threadhelper( new FileOp( ){
                 public void run(JSONArray args) throws FileExistsException, IOException, TypeMismatchException, EncodingException, JSONException {
-                    String dirname=args.getString(0);
-                    /*
-                     * If we don't have the package name in the path, we're reading and writing to places we need permission for
-                     */
-                    if(dirname.contains(cordova.getActivity().getPackageName()) ||
-                            hasReadPermission()) {
-                        String path = args.getString(1);
-                        JSONObject obj = getFile(dirname, path, args.optJSONObject(2), false);
-                        callbackContext.success(obj);
+                    String dirname = args.getString(0);
+                    String path = args.getString(1);
+                    String nativeURL = resolveLocalFileSystemURI(dirname).getString("nativeURL");
+                    boolean containsCreate = args.toString().contains("\"create\":true");
+
+                    if(containsCreate && needPermission(nativeURL, WRITE)) {
+                        getPermissionFile(WRITE);
+                    }
+                    else if(!containsCreate && needPermission(nativeURL, READ)) {
+                        getPermissionFile(READ);
```

Comment:
This code is similar to the `WRITE` case, so it might be better to factor out them both into a function.